File: lists/RemoveDuplicates.java
package lists;

import java.util.HashMap;

public class RemoveDuplicates {

    public static void main(String[] args) {
        SinglyLinkedList l1 = new SinglyLinkedList();
        Node head = new Node("A");
        l1.setHead(head);
        Node n1 = new Node("B");
        head.setNext(n1);
        Node n2 = new Node("B");
        n1.setNext(n2);
        Node n3 = new Node("C");
        n2.setNext(n3);
        Node n4 = new Node("C");
        n3.setNext(n4);
        n4.setNext(new Node("B"));
        printLinkedList(l1);
        removeDuplicatesLoops(l1);
        printLinkedList(l1);
    }

    public static void removeDuplicates(SinglyLinkedList list) {
        HashMap<String,String> map = new HashMap<String,String>();
        Node prior = list.getHead();
        Node next = list.getHead();
        while (next != null) {

            if (!map.containsKey(next.getValue())) {
                map.put(next.getValue(), "Exists");
                prior = next;
                next = next.getNext();
            } else {
                next = next.getNext();
                prior.setNext(next);
            }
        }

    }


    public static void removeDuplicatesLoops(SinglyLinkedList list) {
        Node cur = list.getHead();
        while (cur != null) {
            String val = cur.getValue();
            Node prior = cur;
            Node next = cur.getNext();
            while (next != null) {
                if (next.getValue().equals(val)) {
                    next = next.getNext();
                    prior.setNext(next);
                } else {
                    prior = next;
                    next = next.getNext();
                }
            }
            cur = cur.getNext();
        }
    }


    public static void printLinkedList(SinglyLinkedList list) {
        Node next = list.getHead();
        while (next != null) {
            System.out.print(next.getValue() + " --> ");
            next = next.getNext();
        }
        System.out.println("");
    }

}

File: lists/python/remove_duplicates.py
from LinkedList import LinkedList
from LinkedList import build_ll_from_lst
from Node import Node



def remove_duplicates_sorted(head):
	start = head
	while head is not None:
		while head.next is not None and \
		head.value == head.next.value:
			head.next = head.next.next
		head = head.next
	return LinkedList(start)

def remove_duplicates_unsorted(head):
	if head is None:
		return LinkedList(head)
	hashmap = {}
	hashmap[head.value] = 1
	prior = head
	cur = head.next
	while cur is not None:
		if hashmap.get(cur.value) is not None:
			prior.next = cur.next
		else:
			hashmap[cur.value] = 1
			prior = cur
		cur = cur.next
	return LinkedList(head)



#Tests

def test_remove_duplicates_sorted():
	l1 = build_ll_from_lst([1,2,2,3,3,4])
	l2 = build_ll_from_lst([1,1,1,1,1,1,1])
	l3 = build_ll_from_lst([2,3,3,4,6])
	l4 = build_ll_from_lst([10])

	a1 = build_ll_from_lst([1,2,3,4])
	a2 = build_ll_from_lst([1])
	a3 = build_ll_from_lst([2,3,4,6])
	a4 = build_ll_from_lst([10])

	assert remove_duplicates_sorted(l1.head).lists_eq(a1)
	assert remove_duplicates_sorted(l2.head).lists_eq(a2)
	assert remove_duplicates_sorted(l3.head).lists_eq(a3)
	assert remove_duplicates_sorted(l4.head).lists_eq(a4)

def test_remove_duplicates_unsorted():
	l1 = build_ll_from_lst([1,2,4,2,1,3,7,3,4])
	l2 = build_ll_from_lst([1,1,1,1,1,1,1])
	l3 = build_ll_from_lst([6,2,4,3,3,4,6])
	l4 = build_ll_from_lst([10])

	a1 = build_ll_from_lst([1,2,4,3,7])
	a2 = build_ll_from_lst([1])
	a3 = build_ll_from_lst([6,2,4,3])
	a4 = build_ll_from_lst([10])

	assert remove_duplicates_unsorted(l1.head).lists_eq(a1)
	assert remove_duplicates_unsorted(l2.head).lists_eq(a2)
	assert remove_duplicates_unsorted(l3.head).lists_eq(a3)
	assert remove_duplicates_unsorted(l4.head).lists_eq(a4)

if __name__ == "__main__":
	test_remove_duplicates_sorted()
	test_remove_duplicates_unsorted()
